# Polars (первая часть)

Импорт модуля

In [1]:
import polars as pl

Чтение датасета

In [2]:
df = pl.read_csv('train.csv')

Информация о датасете

In [3]:
df.to_pandas().info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Создание описательных статистик

In [4]:
df.to_pandas().describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Количество пассажиров каждого класса

In [5]:
df.group_by("Pclass").agg(pl.col("PassengerId").count()).sort('Pclass').rename({'PassengerId': 'Count'})

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ Count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 1      ┆ 216   │
│ 2      ┆ 184   │
│ 3      ┆ 491   │
└────────┴───────┘

Количество выживших мужчин и женщин

In [6]:
df.filter(pl.col('Survived') == 1).group_by("Sex").agg(pl.col("PassengerId").count()).rename({'PassengerId': 'Count'})

shape: (2, 2)
┌────────┬───────┐
│ Sex    ┆ Count │
│ ---    ┆ ---   │
│ str    ┆ u32   │
╞════════╪═══════╡
│ male   ┆ 109   │
│ female ┆ 233   │
└────────┴───────┘

Пассажиры старше 44 лет

In [7]:
df.filter(pl.col('Age') > 44)

shape: (115, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0    ┆ null  ┆ S        │
│             ┆          ┆        ┆ (Mary D          ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Kingcome)        ┆   ┆            ┆         ┆       ┆          │
│ 34          ┆ 0        ┆ 2      ┆ Wheadon, Mr.     ┆ … ┆ C.A. 24579 ┆ 10.5    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Edward H         ┆   ┆            ┆         ┆       ┆          │
│ 53          ┆ 1        ┆ 1      ┆ Harper, Mrs.     ┆ … ┆ PC 17572   ┆ 76.7292 ┆ D33   ┆ C        │
│             ┆          ┆        ┆ Henry Sleeper    ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Myna…           ┆   ┆            ┆         ┆       ┆          │
│ …           ┆ …        ┆ …      ┆ …                ┆ … ┆ …          ┆ …       ┆ …     ┆ …        │
│ 858         ┆ 1        ┆ 1      ┆ Daly, Mr. Peter  ┆ … ┆ 113055     ┆ 26.55   ┆ E17   ┆ S        │
│             ┆          ┆        ┆ Denis            ┆   ┆            ┆         ┆       ┆          │
│ 863         ┆ 1        ┆ 1      ┆ Swift, Mrs.      ┆ … ┆ 17466      ┆ 25.9292 ┆ D17   ┆ S        │
│             ┆          ┆        ┆ Frederick Joel   ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Marg…           ┆   ┆            ┆         ┆       ┆          │
│ 872         ┆ 1        ┆ 1      ┆ Beckwith, Mrs.   ┆ … ┆ 11751      ┆ 52.5542 ┆ D35   ┆ S        │
│             ┆          ┆        ┆ Richard Leonard  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (…               ┆   ┆            ┆         ┆       ┆          │
│ 874         ┆ 0        ┆ 3      ┆ Vander Cruyssen, ┆ … ┆ 345765     ┆ 9.0     ┆ null  ┆ S        │
│             ┆          ┆        ┆ Mr. Victor       ┆   ┆            ┆         ┆       ┆          │
│ 880         ┆ 1        ┆ 1      ┆ Potter, Mrs.     ┆ … ┆ 11767      ┆ 83.1583 ┆ C50   ┆ C        │
│             ┆          ┆        ┆ Thomas Jr (Lily  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Ale…             ┆   ┆            ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴──────────────────┴───┴────────────┴─────────┴───────┴──────────┘

# Ускорение работы с Pandas (вторая часть)

Установка пакета

In [116]:
!pip install bottleneck

Импорт библиотек

In [23]:
import numpy as np
import pandas as pd
import bottleneck as bn

Чтение датасета 

In [22]:
df = pd.read_csv('train.csv')

Средний возраст и стандартное отклонение 

In [24]:
print(f'Средний возраст пассажиров\t {bn.nanmean(df.Age)}')
print(f'Стандартное отклонение возраста\t {bn.nanstd(df.Age)}')

Средний возраст пассажиров	 29.69911764705882
Стандартное отклонение возраста	 14.516321150817317


Создание нового столбца Fare_new

In [11]:
np_arr = df[['Fare', 'Pclass']].to_numpy()
new_col = [row[0]*(1.3*(row[1] < 3) + 1.1*(row[1] == 3)) for row in np_arr]
df.insert(10, 'Fare_new', new_col, True)
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare  Fare_new Cabin Embarked  
0        0         A/5 21171   7.2500   7.97500   NaN        S  
1        0          PC 17599  71.2833  92.66829   C85        C  
2        0  STON/O2. 3101282   7.9250   8.71750   NaN        S  
3        0            113803  53.1000  69.03000  C123        S  
4        0            373450   8.0500   8.85500   NaN        S  
..     ...               ...      ...       ...   ...      ...  
886      0            211536  13.0000  16.90000   NaN        S  
887      0            112053  30.0000  39.00000   B42        S  
888      2        W./C. 6607  23.4500  25.79500   NaN        S  
889      0            111369  30.0000  39.00000  C148        C  
890      0            370376   7.7500   8.52500   NaN        Q  

[891 rows x 13 columns]

# Оптимизация типов Pandas (третья часть)

Чтение датасета

In [12]:
df = pd.read_csv('Housing.csv')

Просмотр данных и общей информации о них

In [13]:
df

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

Получение краткой информации

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


Сводка показателей

In [15]:
df.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

Оптимизация использования памяти для различных типов данных:

* Для улучшения использования памяти, рекомендуется категоризировать строковые параметры (например, столбцы mainroad, guestroom, basement, hotwaterheating, airconditioning, prefarea, furnishingstatus).

* Для остальных численных параметров, учитывая их физический смысл, следует использовать беззнаковые типы, так как они не могут иметь отрицательные значения.

* Параметры bedrooms, bathrooms, stories, parking являются целочисленными и ограничены в пределах [0;6], поэтому для них рекомендуется использовать тип uint8.

* Значения параметра area также являются целочисленными и ограничены в пределах [1650;16200], поэтому для их хранения рекомендуется использовать тип uint16.

* Значения параметра price также являются целочисленными и ограничены в пределах [1.75e+6;1.33e+7], поэтому для их хранения рекомендуется использовать тип uint32.

Оптимизация хранения данных

In [16]:
df['price'] = df['price'].astype('uint32')
df['area'] = df['area'].astype('uint16')
df['bedrooms'] = df['bedrooms'].astype('uint8')
df['bathrooms'] = df['bathrooms'].astype('uint8')
df['stories'] = df['stories'].astype('uint8')
df['parking'] = df['parking'].astype('uint8')
df['mainroad'] = df['mainroad'].astype('category')
df['guestroom'] = df['guestroom'].astype('category')
df['basement'] = df['basement'].astype('category')
df['hotwaterheating'] = df['hotwaterheating'].astype('category')
df['airconditioning'] = df['airconditioning'].astype('category')
df['prefarea'] = df['prefarea'].astype('category')
df['furnishingstatus'] = df['furnishingstatus'].astype('category')

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    uint32  
 1   area              545 non-null    uint16  
 2   bedrooms          545 non-null    uint8   
 3   bathrooms         545 non-null    uint8   
 4   stories           545 non-null    uint8   
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    uint8   
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), uint16(1), uint32(1), uint8(4)
memory usage: 10.0 KB


В результате память сократилась в 5.5 раз.